tools/shmsnoop.py

```diff
@@ -0,0 +1,335 @@
+{
+    struct val_t val = {
+        .sys = SYS_SHMGET,
+    };
+    struct pt_regs p;
+
+    if (bpf_probe_read(&p, sizeof(p), (void *) _p))
+        return 0;
+
+    val.key    = PT_REGS_PARM1(&p);
+    val.size   = PT_REGS_PARM2(&p);
+    val.shmflg = PT_REGS_PARM3(&p);
+    return enter(&val);
+};
+
+int trace_shmat(struct pt_regs *ctx, struct pt_regs *_p)
+{
+    struct val_t val = {
+        .sys = SYS_SHMAT,
+    };
+    struct pt_regs p;
+
+    if (bpf_probe_read(&p, sizeof(p), (void *) _p))
+        return 0;
+
+    val.shmid   = PT_REGS_PARM1(&p);
+    val.shmaddr = PT_REGS_PARM2(&p);
+    val.shmflg  = PT_REGS_PARM3(&p);
+    return enter(&val);
+};
+
+int trace_shmdt(struct pt_regs *ctx, struct pt_regs *_p)
+{
+    struct val_t val = {
+        .sys = SYS_SHMDT,
+    };
+    struct pt_regs p;
+
+    if (bpf_probe_read(&p, sizeof(p), (void *) _p))
+        return 0;
+
+    val.shmaddr = PT_REGS_PARM1(&p);
+    return enter(&val);
+};
+
+int trace_shmctl(struct pt_regs *ctx, struct pt_regs *_p)
+{
+    struct val_t val = {
+        .sys = SYS_SHMCTL,
+    };
+    struct pt_regs p;
+
+    if (bpf_probe_read(&p, sizeof(p), (void *) _p))
+        return 0;
+
+    val.shmid = PT_REGS_PARM1(&p);
+    val.cmd   = PT_REGS_PARM2(&p);
+    val.buf   = PT_REGS_PARM3(&p);
+    return enter(&val);
+};
+
+"""
+if args.tid:  # TID trumps PID
+    bpf_text = bpf_text.replace('FILTER',
+        'if (tid != %s) { return 0; }' % args.tid)
+elif args.pid:
+    bpf_text = bpf_text.replace('FILTER',
+        'if (pid != %s) { return 0; }' % args.pid)
+else:
+    bpf_text = bpf_text.replace('FILTER', '')
+
+if debug or args.ebpf:
+    print(bpf_text)
+    if args.ebpf:
+        exit()
+
+# initialize BPF
+b = BPF(text=bpf_text)
+
+syscall_fnname = b.get_syscall_fnname("shmget")
+if BPF.ksymname(syscall_fnname) != -1:
+    b.attach_kprobe(event=syscall_fnname, fn_name="trace_shmget")
+    b.attach_kretprobe(event=syscall_fnname, fn_name="trace_return")
+
+syscall_fnname = b.get_syscall_fnname("shmat")
+if BPF.ksymname(syscall_fnname) != -1:
+    b.attach_kprobe(event=syscall_fnname, fn_name="trace_shmat")
+    b.attach_kretprobe(event=syscall_fnname, fn_name="trace_return")
+
+syscall_fnname = b.get_syscall_fnname("shmdt")
+if BPF.ksymname(syscall_fnname) != -1:
+    b.attach_kprobe(event=syscall_fnname, fn_name="trace_shmdt")
+    b.attach_kretprobe(event=syscall_fnname, fn_name="trace_return")
+
+syscall_fnname = b.get_syscall_fnname("shmctl")
+if BPF.ksymname(syscall_fnname) != -1:
+    b.attach_kprobe(event=syscall_fnname, fn_name="trace_shmctl")
+    b.attach_kretprobe(event=syscall_fnname, fn_name="trace_return")
+
+TASK_COMM_LEN = 16    # linux/sched.h
+
+SYS_SHMGET = 0
+SYS_SHMAT  = 1
+SYS_SHMDT  = 2
+SYS_SHMCTL = 3
+
+initial_ts = 0
+
+class Data(ct.Structure):
+    _fields_ = [
+        ("id",      ct.c_ulonglong),
+        ("ts",      ct.c_ulonglong),
+        ("sys",     ct.c_int),
+        ("key",     ct.c_ulonglong),
```

Comment:
Could you make the type consistent? The C structure is type long and here is ulonglong.